components/tidb_query/src/codec/mysql/binary_literal.rs

```diff
@@ -0,0 +1,466 @@
+    pub fn from_u64(val: u64, byte_size: isize) -> Result<Self> {
+        if byte_size != -1 && (byte_size < 1 || byte_size > 8) {
+            return Err(box_err!("invalid byte size: {}", byte_size));
+        }
+        let mut wtr = vec![];
+        wtr.write_u64::<BigEndian>(val)?;
+        let lit = if byte_size == -1 {
+            Self(trim_leading_zero_bytes(wtr.as_slice()).to_vec())
+        } else {
+            wtr.drain(0..(8 - byte_size) as usize);
+            Self(wtr)
+        };
+        Ok(lit)
+    }
+
+    /// Parses hexadecimal string literal.
+    /// See https://dev.mysql.com/doc/refman/5.7/en/hexadecimal-literals.html
+    pub fn from_hex_str(s: &str) -> Result<Self> {
+        if s.is_empty() {
+            return Err(box_err!(
+                "invalid empty string for parsing hexadecimal literal"
+            ));
+        }
+
+        let trimed = if s.starts_with('x') || s.starts_with('X') {
+            // format is x'val' or X'val'
+            let trimed = s[1..].trim_start_matches('\'');
+            let trimed = trimed.trim_end_matches('\'');
+            if trimed.len() % 2 != 0 {
+                return Err(box_err!(
+                    "invalid hexadecimal format, must even numbers, but {}",
+                    s.len()
+                ));
+            }
+            trimed
+        } else if s.starts_with("0x") {
+            s.trim_start_matches("0x")
+        } else {
+            // here means format is not x'val', X'val' or 0xval.
+            return Err(box_err!("invalid hexadecimal format: {}", s));
+        };
+        if trimed.is_empty() {
+            return Ok(BinaryLiteral(vec![]));
+        }
+        let v = if trimed.len() % 2 != 0 {
+            let mut head = vec![b'0'];
+            head.extend(trimed.as_bytes());
+            box_try!(hex::decode(head))
+        } else {
+            box_try!(hex::decode(trimed.as_bytes()))
+        };
+        Ok(BinaryLiteral(v))
+    }
+
+    /// Parses bit string.
+    /// The string format can be b'val', B'val' or 0bval, val must be 0 or 1.
+    /// See https://dev.mysql.com/doc/refman/5.7/en/bit-value-literals.html
+    pub fn from_bit_str(s: &str) -> Result<Self> {
+        if s.is_empty() {
+            return Err(box_err!("invalid empty string for parsing bit type"));
+        }
+        let trimed = if s.starts_with('b') || s.starts_with('B') {
+            // format is b'val' or B'val'
+            let trimed = s[1..].trim_start_matches('\'');
+            trimed.trim_end_matches('\'')
+        } else if s.starts_with("0b") {
+            s.trim_start_matches("0b")
+        } else {
+            // here means format is not b'val', B'val' or 0bval.
+            return Err(box_err!("invalid hexadecimal format: {}", s));
+        };
+        if trimed.is_empty() {
+            return Ok(BinaryLiteral(vec![]));
+        }
+
+        let aligned_len = (trimed.len() + 7) & (!7);
```

Comment:
Please add some comments explain that the code intends to align the length with factor 8.